### ISSUE TITLE
Edit and Run

### ISSUE THREAD

**Issue by marniks7** [tektoncd/dashboard#2437]
# Is your feature request related to a problem ?

CI can be used not only for automated operations, but to run any task that user might need.

E.g. I would like to run E2E tests (pipeline) manually with some parameters. They can fail.
It is nice to be able to change one (few) parameters and rerun it once again.
Or another operation, e.g. download backup, upload backup or run deploy.

Tekton can be used as common purpose pipeline \ workflow engine, e.g. for Chaos Engineering, and to experiment we need to rerun with different parameters like "duration of chaos", "wait"

# Describe the solution you'd like
I would like to see 'Rerun with Parameters' near 'Rerun' button for TaskRun\PipelineRun\Run

Jenkins analog: https://plugins.jenkins.io/rebuild/

# Describe alternatives you've considered
`kubectl apply -f` with different parameters, no alternatives in UI

# Additional context

- The idea is to use the same form as for `CreatePipelineRun` and add parameter `pipelinerun` to url, this parameter can be used as `template`
`http://localhost:9097/#/pipelineruns/create?namespace=tekton-namespace&pipelinerun=my-previous-pipeline-run`
- New button
![image](https://user-images.githubusercontent.com/9289172/188487812-1a361a03-1979-45e3-b9e6-5998f939339f.png)
- One more problem I see that `Rerun with Parameters` for this approach will support only those parameters that are specified in `CreatePipelineRun.js`, in particular I am worried about absence of workspaces. This can be done in separate PR
- I am willing to implement this functionality if no concerns

**Comment by AlanGreene** [comment]
Thanks for the feature request @marniks7. I think this makes a lot of sense and would be great to add. However, as you mentioned I also have some concerns about missing support in the existing Create PipelineRun UI, particularly for workspaces, but also array params.

Adding a 'Rerun with parameters' (naming might need some work, 'parameters' is an overloaded term) makes it more likely that users will run into issues and could potentially end up with unexpected behaviour depending on how their pipelines are written. I think it's important that the user has full visibility into what they're creating, so not showing unsupported fields but having them passed in the background would not be good. Likewise, stripping unsupported fields wouldn't be great either as it's unlikely this would yield the desired outcome.

We're tracking the missing functionality for the Create UI in https://github.com/tektoncd/dashboard/issues/1283 and https://github.com/tektoncd/dashboard/issues/2171. Contributions on either of those would be very welcome and get us closer to being able to add this new feature safely.

**Comment by marniks7** [comment]
@AlanGreene , I was thinking about this feature and has another proposal:
`Edit YAML and Run` - this is reliable and advanced solution, which will allow to edit YAML, so user will achieve 'whatever floats user boat'.

One more option `Edit and Run` (just renamed Rerun with Parameters)
To make it reliable it will require deep YAML analyze backend vs UI: comparison for all supported fields by UI. In case if all fields supported by UI - show form `Create PipelineRun` with values, otherwise fallback to `Edit YAML and Run`.
However, I have no idea yet how to make such deep analyze :)

**Comment by AlanGreene** [comment]
A YAML input to create a run or edit and rerun an existing one sounds good to me as an escape hatch for unsupported features in the UI or if someone just has a blob of YAML somewhere already that they want to execute.

> To make it reliable it will require deep YAML analyze backend 

Can you elaborate on what you mean by this?

**Comment by marniks7** [comment]
`unsupported features in the UI` are inevitable, we cannot keep up with all backend alpha features and different backend \ UI versions can be used. 
Nevertheless, in order to support `Edit and Run` (without YAML as input), we need to be able to understand if ALL fields in current yaml are supported by current UI or not.
For example, list of supported fields can be described as those:
- `apiVersion`
- `kind`
- `spec.workspaces.[].name`
- `spec.workspaces.[].configMap.name`
- `spec.workspaces.[].secret.secretName`
- `spec.pipelineRef.name`
- `spec.serviceAccountName`

If PipelineRun YAML contains any field except above - we fallback to direct YAML edit, otherwise `Edit and Run`

**Comment by AlanGreene** [comment]
I think a good starting point would be to always present the option to switch to 'YAML mode'. That would unblock users who want to use features not currently supported in the UI, and buy us time to figure out the best approach to handle this (new UI, comparison of supported features, detect when UI doesn't support a used feature, etc.).

**Comment by marniks7** [comment]
Hi @AlanGreene,
Could you please check the solution. If no objections - I can implement it.

### Create PipelineRun (TaskRun, Run) as YAML
1. Generate models for Tekton Run(s) from CRDs or OpenAPI schema using [kubernetes-models-ts](https://github.com/tommy351/kubernetes-models-ts)
2. Leave only fields that are supported by UI?!
3. Change `CreatePipelineRun.js` to use those models
4. Use [Monaco-Editor](https://github.com/microsoft/monaco-editor) and add [React Monaco Editor](https://github.com/react-monaco-editor/react-monaco-editor)
5. On CreatePipelineRun page add toggle to switch to yaml. **Could you please add your vision where this toggle (or maybe another element type) should be?**
Toggle:
![image](https://user-images.githubusercontent.com/9289172/193493521-c0863a75-5249-46ff-834b-16b38067a4f6.png)
Monaco-Editor:
![image](https://user-images.githubusercontent.com/9289172/193493670-c79b6f95-9721-408c-ba27-464c6ae6763d.png)
7. In order to switch from model to yaml just convert to YAML
8. In order to switch from yaml to model back generate jsonMergePatch between json model before switch and latest YAML (converted to json). Try to apply this patch to the model. in case if failed prohibit switch to model. Uses [json-merge-patch](https://github.com/pierreinglebert/json-merge-patch)
9. Add support for `yamlmode=true` to the query params when yaml mode is turned on

### Edit and Run
1. Add `pipelinerun` to the URL query params
2. Get pipelinerun
3. Generate jsonMergePatch and try to apply to model. if failed - switch to yamlmode

### Important Notes
1. Unfortunately, Content-Security-Policy will require downgrade because of
https://github.com/microsoft/monaco-editor/issues/271
2. Possible improvement. For yaml-schema suggestions & validations use [monaco-yaml](https://github.com/remcohaszing/monaco-yaml). Because monaco doesn't support it for `yaml` https://github.com/microsoft/monaco-editor/issues/1470

**Comment by AlanGreene** [comment]
/remove-lifecycle stale
This is currently in progress. We already have a YAML mode for creating a PipelineRun, and there's a PR open adding the ability to edit YAML of an existing run as the basis for creating a new one.

**Comment by AlanGreene** [comment]
We recently updated the Dashboard to display list and details pages for the new v1beta1 CustomRun (replacement for the deprecated v1alpha1 Run).

Now that we have the YAML mode for Create PipelineRun we should add the same (including this edit and rerun feature) for both TaskRun and CustomRun to ensure consistency.

**Comment by AlanGreene** [comment]
Edit and run functionality is available for PipelineRun, TaskRun, and CustomRun resources via the YAML mode editor. There are no current plans to expand the form-based UI capabilities so closing for now. We may revisit this in future as part of broader changes to the app.
